flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java

```diff
@@ -165,20 +286,36 @@ public void beforeBulk(long executionId, BulkRequest request) { }
 				@Override
 				public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
 					if (response.hasFailures()) {
-						for (BulkItemResponse itemResp : response.getItems()) {
-							Throwable failure = callBridge.extractFailureCauseFromBulkItemResponse(itemResp);
+						BulkItemResponse itemResponse;
+						Throwable failure;
+
+						for (int i = 0; i < response.getItems().length; i++) {
+							itemResponse = response.getItems()[i];
+							failure = callBridge.extractFailureCauseFromBulkItemResponse(itemResponse);
 							if (failure != null) {
-								LOG.error("Failed Elasticsearch item request: {}", failure.getMessage(), failure);
-								failureThrowable.compareAndSet(null, failure);
+								LOG.error("Failed Elasticsearch item request: {}", itemResponse.getFailureMessage(), failure);
+
+								if (failureHandler.onFailure(request.requests().get(i), failure, requestIndexer)) {
```

Comment:
I wonder if it would be better if the `onFailure` method would not return a boolean but throw a Throwable?
This way users have more flexibility in implementing their failure handler.

For example if a failure handler is doing three retries and fails afterwards, the original exception will be thrown. If the `onFailure()` method can throw their own exception, you can throw a custom exception that tells the user about the three retries.

We can definitively discuss this because this change is annoying to do (docs & javadocs need to be updated).